src/node/txorphanage.cpp

```diff
@@ -83,373 +216,484 @@ class TxOrphanageImpl final : public TxOrphanage {
-    // have been mined or received.
-    // 100 orphans, each of which is at most 100,000 bytes big is
-    // at most 10 megabytes of orphans and somewhat more byprev index (in the worst case):
-    unsigned int sz = GetTransactionWeight(*tx);
-    if (sz > MAX_STANDARD_TX_WEIGHT)
-    {
-        LogDebug(BCLog::TXPACKAGES, "ignoring large orphan tx (size: %u, txid: %s, wtxid: %s)\n", sz, hash.ToString(), wtxid.ToString());
+bool TxOrphanageImpl::IsUnique(Iter<ByWtxid> it) const
+{
+    // Iterators ByWtxid are sorted by wtxid, so check if neighboring elements have the same wtxid.
+    auto& index = m_orphans.get<ByWtxid>();
+    if (it == index.end()) return false;
+    if (std::next(it) != index.end() && std::next(it)->m_tx->GetWitnessHash() == it->m_tx->GetWitnessHash()) return false;
+    if (it != index.begin() && std::prev(it)->m_tx->GetWitnessHash() == it->m_tx->GetWitnessHash()) return false;
+    return true;
+}
+
+TxOrphanage::Usage TxOrphanageImpl::UsageByPeer(NodeId peer) const
+{
+    auto it = m_peer_orphanage_info.find(peer);
+    return it == m_peer_orphanage_info.end() ? 0 : it->second.m_total_usage;
+}
+
+TxOrphanage::Count TxOrphanageImpl::CountAnnouncements() const { return m_orphans.size(); }
+
+TxOrphanage::Usage TxOrphanageImpl::TotalOrphanUsage() const { return m_unique_orphan_bytes; }
+
+TxOrphanage::Count TxOrphanageImpl::CountUniqueOrphans() const { return m_unique_orphans; }
+
+TxOrphanage::Count TxOrphanageImpl::AnnouncementsFromPeer(NodeId peer) const {
+    auto it = m_peer_orphanage_info.find(peer);
+    return it == m_peer_orphanage_info.end() ? 0 : it->second.m_count_announcements;
+}
+
+TxOrphanage::Count TxOrphanageImpl::LatencyScoreFromPeer(NodeId peer) const {
+    auto it = m_peer_orphanage_info.find(peer);
+    return it == m_peer_orphanage_info.end() ? 0 : it->second.m_total_latency_score;
+}
+
+bool TxOrphanageImpl::AddTx(const CTransactionRef& tx, NodeId peer)
+{
+    const auto& wtxid{tx->GetWitnessHash()};
+    const auto& txid{tx->GetHash()};
+
+    // Ignore transactions above max standard size to avoid a send-big-orphans memory exhaustion attack.
+    TxOrphanage::Usage sz = GetTransactionWeight(*tx);
+    if (sz > MAX_STANDARD_TX_WEIGHT) {
+        LogDebug(BCLog::TXPACKAGES, "ignoring large orphan tx (size: %u, txid: %s, wtxid: %s)\n", sz, txid.ToString(), wtxid.ToString());
         return false;
     }
 
-    auto ret = m_orphans.emplace(wtxid, OrphanTx{{tx, {peer}}, Now<NodeSeconds>() + ORPHAN_TX_EXPIRE_TIME, m_orphan_list.size()});
-    assert(ret.second);
-    m_orphan_list.push_back(ret.first);
-    for (const CTxIn& txin : tx->vin) {
-        m_outpoint_to_orphan_it[txin.prevout].insert(ret.first);
-    }
-    m_total_orphan_usage += sz;
-    m_total_announcements += 1;
+    // We will return false if the tx already exists under a different peer.
+    const bool brand_new{!HaveTx(wtxid)};
+
+    auto ret = m_orphans.get<ByWtxid>().emplace(tx, peer, m_current_sequence);
```

Comment:
in d5097d6eef85eab971352cf78fb94a30c6e6f127: nit: could use a structured binding for getting rid of many `.first/.second` accesses (that I often find confusing) and enhancing the readability, e.g.:
```suggestion
    auto [new_announcement_it, inserted] = m_orphans.get<ByWtxid>().emplace(tx, peer, m_current_sequence);
```
(here and in many other return values of `.emplace` calls)

Reply:
Nice, changed